### ISSUE TITLE
504 for prod for 3D Mitochondria for FruitFly

### ISSUE THREAD

**Issue by inghylt** [MetabolicAtlas/MetabolicAtlas#1200]
**Describe the bug**

https://metatlas-prod3.csbi.chalmers.se/api/v2/3d-network?model=FruitflyGem&version=1_2_0&type=compartment&id=mitochondria

https://metatlas-prod3.csbi.chalmers.se/api/v2/3d-network?model=YeastGem&version=8_6_2&type=compartment&id=cytoplasm

These two request fail

**Steps to reproduce**

Visit the links

**Expected behavior**

- [ ] That the requests don't fail
- [ ] If applicable: Tests verifying the fix are added

**Screenshots**

**Additional context**

**Comment by e0** [comment]
Tested these on the dev server now. The first one loaded on the first try. The second one timeout initially but managed to load on the second try.

Weird that the dev server seems to be performing better than the new prod server.

Also.. I don’t think we are allocating the resources in a smart enough way. We have the  `0.40` cap on `cpus` for the `api` container, I noticed that the rest of the system was basically resting when `api` was at `0.40`. Would be more ideal if it can use more compute power if they are available. It should speed up the response time as well for these bigger maps.

**Comment by e0** [comment]
According to [the docs](https://docs.docker.com/config/containers/resource_constraints/#cpu), the default behavior is that each container’s access to the host machine’s CPU cycles is unlimited. It seems like if we do not set any hard caps for any containers then the system will try to manage everything by default.

I will try this out on the dev server. Please let me know if this has been tried before and the hard caps are absolutely needed, and if so why.

**Comment by e0** [comment]
Have deployed without setting `cpus` for anything to the dev server now. Both of the endpoints above loaded in just under `25s`. I have tested some other pages and I think everything seems faster than `prod`.

**Comment by inghylt** [comment]
Nice! I'm wondering though if we should set some kind of limit so the api does not risk crippling the rest of the containers on the machine? It would be nice to stress test things to make sure this is never the case

**Comment by mihai-sysbio** [comment]
> the api does not risk crippling the rest of the containers on the machine

This is precisely the goal of allocating CPU resources.

> much more ideal and efficient to let the system manage its resources instead of us capping things manually.

What we would like to avoid is say the Neo4j container taking up so much compute that it slows down other requests that go via the API to Postgres. While I see how it may seem that it is better to avoid resource _capping_, what other way is there to enforce that for example Nginx should _always_ have priority over other containers?

**Comment by inghylt** [comment]
We talked a bit about this during standup. I've done some research and I think the option [cpu-shares](https://docs.docker.com/config/containers/resource_constraints/) sounds like the best solution

_"This is only enforced when CPU cycles are constrained. When plenty of CPU cycles are available, all containers use as much CPU as they need. In that way, this is a soft limit."_

Note though that _"It prioritizes container CPU resources for the available CPU cycles. It does not guarantee or reserve any specific CPU access."_ So I guess setting `cpu_share` would not imply 100% guarantee that nginx will never be crippled. [This article](https://www.batey.info/cgroup-cpu-shares-for-docker.html) says _"cpu_shares dictate how much CPU time each container has access to."_, note the word dictates which seems more reassuring, though I would trust the offical docs more :upside_down_face: I still believe this is the best option for us since we are not worsening the performance by capping CPU access when it's uncalled-for, but we are able to always prioritize nginx CPU usage before the other containers.

A container’s `cpu_share` is a relative value and the default value for a container for this is `1024`. Thus I would suggest a value of `2048` for nginx so it has double the priority compared to the other containers when CPU cycles are constrained.

Thoughts on this? @MalinAhlberg @mihai-sysbio @nanjiangshu @e0

**Comment by e0** [comment]
Great research @inghylt!

Regarding the blocking issue. What I have seen is that it is already a problem. For example if a user is requesting a large network, then all neo4j requests could be blocked while the network is being loaded. In theory it is good if this does not block other things like postgres. In practice, I would probably still favor processing all requests as fast as possible by not capping any resources, which should minimize overall blocked time.

I do think a stress test is needed to help conclude this issue. If we find that the performance gains outweigh the inter-containers blocking issue, and that those issues are not crucial, then I would recommend not capping any resources.

**Comment by inghylt** [comment]
@e0 just to be clear, you are suggesting that

1) we should have no configuration in terms of CPU at all
2) we use the `cpu_share`? (Which is not really a cap)
3) something else

**Comment by e0** [comment]
I'm suggesting 1.

It's good that @mihai-sysbio mentioned that scenario. That should be something that we focus on in the stress test. It would be most ideal if we can allocate like max 90% of the cpu for all non-nginx containers but I don't know if that's possible.

**Comment by nanjiangshu** [comment]
I think a better solution is to ask the data center to provide a more powerful production server, it should be easy to get a VM with 16 CPU cores + 32 GB RAM. Temporary I can allocate such a virtual machine from SNIC cloud since it may take some time for the data center to do the work.

**Comment by nanjiangshu** [comment]
Now I think the problem might be caused Cloudfare setting, when I tried with https://dev-metatlas.bioshu.se/api/v2/3d-network?model=FruitflyGem&version=1_2_0&type=compartment&id=mitochondria. That VM machine has 8 CPU cores and 16GB RAM. The CPUs are also constrained in the same way as for the production server, i.e. 0.4 for the api and the CPU is much worse compared to Intel Xeon Gold 6230 @ 2.10GHz. But the same API request returned result in a reasonable time.

**Comment by inghylt** [comment]
@nanjiangshu did you redeploy to dev? Because if not, what's on dev right now is what @e0 deployed as described here https://github.com/MetabolicAtlas/MetabolicAtlas/issues/1200#issuecomment-1308806028, i.e. no constraints on cpu

**Comment by nanjiangshu** [comment]
@inghylt The link I posted is not the dev server `dev-metatlas.csbi.chalmers.se` but my own server, please note the ending of the domain name is `bioshu.se`

**Comment by mihai-sysbio** [comment]
> Is the limit set to 60 seconds? Where does it set?

I was expecting the timeout to be 30s according to https://github.com/MetabolicAtlas/MetabolicAtlas/blob/d311af7e3dc4367189e6f2b67c574dfe6ef2412b/nginx/prod.nginx.conf#L6

**Comment by e0** [comment]
Thanks for setting up the new dev server!

I did a bit of testing now on the dev server now, which currently has no manual `cpu` allocation and has caching disabled.

The two questions I had in mind were:

1. During stress testing, if one container is fully loaded, does it block other containers?
2. During regular usage, how are the response times?

### 1. Stress testing

For stress testing, I used `locust`, which includes the two problematic endpoints, as well as the `/api/v2/repository/integrated_models` one. During this, the problematic endpoints would consistently fail, and the `node` process would get to `299% CPU` (out of `300%` or 3 cores). However the `integrated_models` endpoint never failed a single time.

I also manually browsed the website, visiting the gem browser, map viewer, interaction partners pages etc. These also loaded without any problems. I specifically tested out the GotEnzymes pages as well and I noticed how the cpu usage for `node` would go down to `~250%` as `~35%` was allocated to `postgres`. It seems like the system is really smart about not blocking itself and distributing resources.

Based on this, the conclusions are:
- Slow endpoints that are requested at around the same time will likely fail.
- Fast endpoints load as usual under heavy server load.

### 2. Regular usage

Since caching is normally enabled, it seems unlikely that multiple slow endpoints would be requested for the first time in a small time window. So I waited until the server calmed down (no active requests being processed, `node` at around `0.3% CPU`) before testing the two problematic endpoints individually.

For: `/api/v2/3d-network?model=FruitflyGem&version=1_2_0&type=compartment&id=mitochondria`, the load time is around `22-24s`.

For: `/api/v2/3d-network?model=YeastGem&version=8_6_2&type=compartment&id=cytoplasm`, the load time is around `23-25s`.

Each of these requests seem to run the `node` process on a single node, so the CPU usage only goes up to `99%`, which means `~200%` is still available for other things. I also tried loading both simultaneously and as expected the CPU usage went up to `199%` and the load times are within the times listed above.

Based on this, the conclusions are:

- The problematic endpoints should load within the timeout `30s` under regular usage and server state (no caching).
- Unless 3 or more slow endpoints are being requested at the same time, the server should have free CPU ready to be allocated for other things.

### Summary

Combining the results for these two scenarios (stress testing and regular usage), it seems like the server would be more efficient at handling requests. In the unlikely event of multiple (3 or more) slow endpoints being requested at the same time, the server should still be able to handle and load other requests. Any problematic endpoint should also manage to load and be added to the cache once the server is under less stress. This is the exact behavior of the production server at the moment, only that it's more likely for the production server to fail given the `0.4` CPU constraint.

Since there doesn't seem to be any downside and the gain in load times are quite significant, my current recommendation is that we move forward with the manual `cpu` allocation removed.

**Comment by mihai-sysbio** [comment]
Very nice results 💪🏻  clearly communicated 👍🏻  

> the `integrated_models` endpoint never failed a single time.

Just a clarification, if I remember correctly this request is a json file from disk, so no computation is needed, and then it's cached. That's a good thing, since if this request would fail the website would throw an error on the homepage.

**Comment by inghylt** [comment]
@e0 Great work and summary, very nice to see that the website is always available despite a heavy load on the api/other containers! Looking forward to an even snappier performance of the site after #1217 is merged and deployed

**Comment by nanjiangshu** [comment]
Great analysis @e0! I think it is clear enough that the gain of removing the CPU constrain is more than the loss. My general impression with Linux is that the system does a great job with CPU management but less satisfactory with memory.

**Comment by inghylt** [comment]
@e0 It would be nice if we could store the locust file you used somewhere so it can be reused in the future!

**Comment by e0** [comment]
That's a good idea! A lot of the older ones are not functional anymore actually with the route changes. We should probably update it at some point. I can paste what I used here for now in case we want to reuse it.

```python
from locust import HttpUser, task, between

class RandomUser(HttpUser):
    wait_time = between(1, 10)

    @task(1)
    def index(self):
        self.client.get("/api/v2/repository/integrated_models/")

    @task(2)
    def explore(self):
        self.client.get("/api/v2/3d-network?model=FruitflyGem&version=1_2_0&type=compartment&id=mitochondria")

    @task(3)
    def explore(self):
        self.client.get("/api/v2/3d-network?model=YeastGem&version=8_6_2&type=compartment&id=cytoplasm")

    def on_start(self):
        self.client.get("/")
```

**Comment by mihai-sysbio** [comment]
@e0 I think @inghylt was hinting at updating the [existing but outdated locustfile](https://github.com/MetabolicAtlas/MetabolicAtlas/blob/main/locustfile.py) with a config that even though not perfect at least functional at the moment of testing

**Comment by e0** [comment]
> @e0 I think @inghylt was hinting at updating the [existing but outdated locustfile](https://github.com/MetabolicAtlas/MetabolicAtlas/blob/main/locustfile.py) with a config that even though not perfect at least functional at the moment of testing

That was the same file I modified locally. I would much rather that the team has a discussion about this and then create a new issue if we want to work with this. The main reasons why I didn't want to include this change in the repo were:

1. The file includes many endpoints which are outdated and do not work anymore.
2. Only including the ones as posted [here](https://github.com/MetabolicAtlas/MetabolicAtlas/issues/1200#issuecomment-1315467400) gives really poor coverage of our API.
3. Since we decided that JavaScript should be the default language when working with this project, we should probably get rid of `locust` entirely and use something like [k6](https://github.com/grafana/k6).
